test/common/request_info/dynamic_metadata_impl_test.cc

```diff
@@ -0,0 +1,165 @@
+    }
+  }
+
+  int access() const {
+    if (access_count_) {
+      ++*access_count_;
+    }
+    return value_;
+  }
+
+private:
+  int value_;
+  size_t* access_count_;
+  size_t* destruction_count_;
+};
+
+class SimpleType : public DynamicMetadata::Object {
+public:
+  SimpleType(int value) : value_(value) {}
+
+  int access() const { return value_; }
+
+private:
+  int value_;
+};
+
+class DynamicMetadataImplTest : public testing::Test {
+public:
+  DynamicMetadataImplTest() { resetDynamicMetadata(); }
+
+  void resetDynamicMetadata() { dynamic_metadata_ = std::make_unique<DynamicMetadataImpl>(); }
+  DynamicMetadata& dynamic_metadata() { return *dynamic_metadata_; }
+
+private:
+  std::unique_ptr<DynamicMetadataImpl> dynamic_metadata_;
+};
+
+} // namespace
+
+TEST_F(DynamicMetadataImplTest, Simple) {
+  size_t access_count = 0u;
+  size_t destruction_count = 0u;
+  dynamic_metadata().setData(
+      "test_name", std::make_unique<TestStoredTypeTracking>(5, &access_count, &destruction_count));
+  EXPECT_EQ(0u, access_count);
+  EXPECT_EQ(0u, destruction_count);
+
+  EXPECT_EQ(5, dynamic_metadata().getData<TestStoredTypeTracking>("test_name").access());
+  EXPECT_EQ(1u, access_count);
+  EXPECT_EQ(0u, destruction_count);
+
+  resetDynamicMetadata();
+  EXPECT_EQ(1u, access_count);
+  EXPECT_EQ(1u, destruction_count);
+}
+
+TEST_F(DynamicMetadataImplTest, SameTypes) {
+  size_t access_count_1 = 0u;
+  size_t access_count_2 = 0u;
+  size_t destruction_count = 0u;
+  static const int ValueOne = 5;
+  static const int ValueTwo = 6;
+
+  dynamic_metadata().setData("test_1", std::make_unique<TestStoredTypeTracking>(
+                                           ValueOne, &access_count_1, &destruction_count));
+  dynamic_metadata().setData("test_2", std::make_unique<TestStoredTypeTracking>(
+                                           ValueTwo, &access_count_2, &destruction_count));
+  EXPECT_EQ(0u, access_count_1);
+  EXPECT_EQ(0u, access_count_2);
+  EXPECT_EQ(0u, destruction_count);
+
+  EXPECT_EQ(ValueOne, dynamic_metadata().getData<TestStoredTypeTracking>("test_1").access());
+  EXPECT_EQ(1u, access_count_1);
+  EXPECT_EQ(0u, access_count_2);
+  EXPECT_EQ(ValueTwo, dynamic_metadata().getData<TestStoredTypeTracking>("test_2").access());
+  EXPECT_EQ(1u, access_count_1);
+  EXPECT_EQ(1u, access_count_2);
+  resetDynamicMetadata();
+  EXPECT_EQ(2u, destruction_count);
+}
+
+TEST_F(DynamicMetadataImplTest, SimpleType) {
+  dynamic_metadata().setData("test_1", std::make_unique<SimpleType>(1));
+  dynamic_metadata().setData("test_2", std::make_unique<SimpleType>(2));
+
+  EXPECT_EQ(1, dynamic_metadata().getData<SimpleType>("test_1").access());
+  EXPECT_EQ(2, dynamic_metadata().getData<SimpleType>("test_2").access());
+}
+
+TEST_F(DynamicMetadataImplTest, NameConflict) {
+  dynamic_metadata().setData("test_1", std::make_unique<SimpleType>(1));
+  EXPECT_THROW(dynamic_metadata().setData("test_1", std::make_unique<SimpleType>(2)),
```

Comment:
Can we do EXPECT_THROW_WITH_MESSAGE for here and below?

Reply:
Done.